runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkRuntimeContext.java

```diff
@@ -118,6 +117,24 @@ public CoderRegistry getCoderRegistry() {
     return coderRegistry;
   }
 
+  private static class PipelineOptionsHolder {
+    // on executors, this should deserialize once.
+    private static transient volatile PipelineOptions pipelineOptions = null;
+
+    static PipelineOptions getOrInit(String serializedPipelineOptions) {
+      if (pipelineOptions == null) {
+        synchronized (PipelineOptionsHolder.class) {
+          if (pipelineOptions == null) {
+            pipelineOptions = deserializePipelineOptions(serializedPipelineOptions);
+          }
+        }
+        // register IO factories.
+        IOChannelUtils.registerIOFactoriesAllowOverride(pipelineOptions);
```

Comment:
Comment after the merge:

Can it happen, in some clusters, that this process on the worker runs multiple pipelines at the same time? If so, different pipelines may have different pipeline options, and the second registration may clobber the first one.

Is this relevant? If not, great. If so, perhaps we should log a JIRA issue for later.